### PR TITLE
backup: control backup threads

### PR DESCRIPTION
**Pull request by NingLin-P**
Signed-off-by: linning <linningde25@gmail.com>

###  What have you changed?

control backup endpoint's thread pool

###  What is the type of the changes?

- Improvement (a change which is an improvement to an existing feature)

###  How is the PR tested?

No

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

No

###  Does this PR affect `tidb-ansible`?

No

### CODE REVIEW

File: components/backup/src/endpoint.rs [hunk 1]
```diff
@@ -267,21 +276,76 @@ impl<R: RegionInfoProvider> Progress<R> {
...
+        if self.size >= new_size && self.size - new_size <= 10 {
+            return;
+        }
         let workers = ThreadPoolBuilder::new()
             .name_prefix("backup-worker")
-            .pool_size(8) // TODO: make it configure.
+            .pool_size(new_size)
             .build();
+        let _ = self.workers.replace(workers);
+        self.size = new_size;
```
Comment (overvenus) [inline components/backup/src/endpoint.rs]: Could you add a metric to record the current pool size?

File: components/backup/src/endpoint.rs [hunk 2]
```diff
@@ -267,21 +276,76 @@ impl<R: RegionInfoProvider> Progress<R> {
...
+        F: Future<Item = (), Error = ()> + Send + 'static,
+    {
+        self.workers.as_ref().unwrap().spawn(future);
+    }
+
+    /// Lazily adjust the thread pool's size
+    ///
+    /// Resizing if the thread pool need to expend or there
+    /// are too many idle threads. Otherwise do nothing.
+    fn adjust_with(&mut self, new_size: usize) {
```
Comment (overvenus) [inline components/backup/src/endpoint.rs]: Could you test the resize feature?

File: components/backup/src/endpoint.rs [hunk 3]
```diff
@@ -18,25 +21,30 @@ use tikv::raftstore::store::util::find_peer;
...
 use crate::*;
 
 const WORKER_TAKE_RANGE: usize = 6;
-const WORKER_NUM: usize = 8;
+
+// used to periodically check whether the thread pool are idle.
+const IDLE_THREADPOOL_CHECK_INTERVAL: u64 = 30_000; // 30 secs
+
+// if thread pool has been idle for such long time, we will shutdown it.
+const IDLE_THREADPOOL_DURATION: u64 = 30_000; // 30 secs
```
Comment (overvenus) [inline components/backup/src/endpoint.rs]: Could you unify the unit of these two constants?

File: components/backup/src/endpoint.rs [hunk 4]
```diff
@@ -438,6 +505,13 @@ impl<E: Engine, R: RegionInfoProvider> Runnable<Task> for Endpoint<E, R> {
     }
 }
 
+impl<E: Engine, R: RegionInfoProvider> RunnableWithTimer<Task, ()> for Endpoint<E, R> {
+    fn on_timeout(&mut self, timer: &mut Timer<()>, _: ()) {
+        self.pool.borrow_mut().check_active();
+        timer.add_task(Duration::from_millis(IDLE_THREADPOOL_CHECK_INTERVAL), ());
```
Comment (overvenus) [inline components/backup/src/endpoint.rs]: I'd like to extend the interval to avoid create and destroy threads too frequently.

Comment (NingLin-P) [reply components/backup/src/endpoint.rs]: this interval is for periodically check whether thread pool is idle, we can change another constant IDLE_THREADPOOL_DURATION to do this. Btw, how much time you think is proper.

Comment (overvenus) [reply components/backup/src/endpoint.rs]: I think 30min should be fine.

Comment (NingLin-P) [reply components/backup/src/endpoint.rs]: OK

File: components/backup/src/endpoint.rs [hunk 5]
```diff
@@ -438,6 +503,13 @@ impl<E: Engine, R: RegionInfoProvider> Runnable<Task> for Endpoint<E, R> {
     }
 }
 
+impl<E: Engine, R: RegionInfoProvider> RunnableWithTimer<Task, ()> for Endpoint<E, R> {
+    fn on_timeout(&mut self, timer: &mut Timer<()>, _: ()) {
+        self.pool.borrow_mut().check_active();
```
Comment (overvenus) [inline components/backup/src/endpoint.rs]: Could you test whether the pool can be shutdown?

Comment (NingLin-P) [reply components/backup/src/endpoint.rs]: Sure.

File: components/backup/src/metrics.rs [hunk 6]
```diff
@@ -22,6 +22,11 @@ lazy_static! {
         exponential_buckets(32.0, 2.0, 20).unwrap()
     )
     .unwrap();
+    pub static ref BACKUP_THREAD_POOL_SIZE_GAUGE: Gauge = register_gauge!(
```
Comment (5kbpers) [inline components/backup/src/metrics.rs]: How about `IntGauge`?

Comment (NingLin-P) [reply components/backup/src/metrics.rs]: Yes, it is more suitable than `Gauge`.

**Pull request merged**